src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java

```diff
@@ -0,0 +1,240 @@
+import htsjdk.samtools.util.StringUtil;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.api.java.function.Function;
+import org.broadinstitute.hellbender.cmdline.Argument;
+import org.broadinstitute.hellbender.cmdline.CommandLineProgramProperties;
+import org.broadinstitute.hellbender.cmdline.StandardArgumentDefinitions;
+import org.broadinstitute.hellbender.cmdline.programgroups.SparkProgramGroup;
+import org.broadinstitute.hellbender.engine.spark.GATKSparkTool;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.tools.picard.analysis.BaseDistributionByCycleMetrics;
+import org.broadinstitute.hellbender.tools.picard.analysis.CollectBaseDistributionByCycle;
+import org.broadinstitute.hellbender.utils.R.RScriptExecutor;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.gcs.BucketUtils;
+import org.broadinstitute.hellbender.utils.io.Resource;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+import java.io.*;
+import java.util.Arrays;
+import java.util.List;
+
+@CommandLineProgramProperties(
+        summary = "Program to chart the nucleotide distribution per cycle in a SAM or BAM file.",
+        oneLineSummary = "Program to chart the nucleotide distribution per cycle in a SAM or BAM file.",
+        programGroup = SparkProgramGroup.class
+)
+public final class CollectBaseDistributionByCycleSpark extends GATKSparkTool {
+
+    private static final long serialVersionUID = 1L;
+
+    @Argument(doc = "uri for the output file: a local file path",
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            optional = true)
+    public String out;
+
+    @Argument(shortName="C", fullName = "chart", doc="A file (with .pdf extension) to write the chart to.", optional=true)
+    public File chartOutput;
+
+    @Argument(shortName="A", fullName = "alignedReadsOnly", doc="If set to true, calculate the base distribution over aligned reads only.")
+    public boolean alignedReadsOnly = false;
+
+    @Argument(shortName="F", fullName = "pfReadsOnly", doc="If set to true calculate the base distribution over PF reads only.")
+    public boolean pfReadsOnly = false;
+
+    private static final class HistogramGenerator implements Serializable{
+        private static final long serialVersionUID = 1L;
+
+        private static final int NBASES = 5; //ACGT + unknown
+        private int maxLengthSoFar = 0;
+        private final long[][] firstReadTotalsByCycle = new long[NBASES][maxLengthSoFar];
+        private long[] firstReadCountsByCycle = new long[maxLengthSoFar];
+        private final long[][] secondReadTotalsByCycle = new long[NBASES][maxLengthSoFar];
+        private long[] secondReadCountsByCycle = new long[maxLengthSoFar];
+        private boolean seenSecondEnd = false;
```

Comment:
or maybe seenAnySecondOfPair (because this appears to only be used as a signal that there is any data from the second read of pairs).